File: apps/api/src/lunaris_api/routers/capabilities.py
from fastapi import APIRouter

from ..capabilities import CapabilityStatus, resolve_capabilities
from ..dependencies import CredentialVaultDep, OptionalUserIdDep, SecretStoreDep
from ..schemas import CapabilityStatusView

router = APIRouter(prefix="/api/capabilities", tags=["capabilities"])


def _to_view(status_: CapabilityStatus) -> CapabilityStatusView:
    return CapabilityStatusView(
        capability=status_.capability, mode=status_.mode, provider=status_.provider
    )


@router.get("", response_model=list[CapabilityStatusView])
async def get_capabilities(
    owner_id: OptionalUserIdDep,
    vault: CredentialVaultDep,
    store: SecretStoreDep,
) -> list[CapabilityStatusView]:
    """Which provider each key-gated capability is using RIGHT NOW: its keyed provider (live) or its
    keyless local fallback. The web shows a Draft badge per fallback capability; a capability flips
    to live the moment its key is stored.

    Tenant-aware: with BYOK on and an authenticated caller, it reads the caller's own credential
    presence (the keys they've set); otherwise it reads the process-wide file secret store
    (single-user / admin). Only set/unset is read — never a key value.
    """
    if vault is not None and owner_id is not None:
        statuses = await vault.statuses(user_id=owner_id)
        set_names = {s.provider for s in statuses if s.is_set}
    else:
        set_names = {s.name for s in store.statuses() if s.is_set}
    return [_to_view(c) for c in resolve_capabilities(lambda name: name in set_names)]

File: apps/web/src/lib/capabilities.ts
import type { CapabilityMode, CapabilityName } from "../types/course";
import { authedFetch } from "./apiClient";

/** Which provider a key-gated capability is using right now: its keyed provider ("live") or its
 *  keyless local fallback. Flips to "live" the moment the capability's key is stored. Tenant-aware
 *  under BYOK (reads the caller's own keys). */
export interface CapabilityStatus {
  capability: CapabilityName;
  mode: CapabilityMode;
  provider: string;
  /** For a keyless fallback that runs on the local model server (the LLM): whether inference is on
   *  GPU or CPU. Absent/null for live capabilities and keyless web services (search/video). */
  compute?: "cpu" | "gpu" | null;
}

/** Human label per capability — shared by the live settings badge and the per-course build tag so
 *  the two indicators name a capability identically. */
export const CAPABILITY_LABELS: Record<CapabilityName, string> = {
  llm: "Language model",
  embeddings: "Embeddings",
  search: "Web search",
  video: "Video",
  cover: "Image generation",
};

/** Fetch the per-capability live/fallback status. Best-effort: a failure resolves to an empty list
 *  (the Draft banner simply doesn't show) rather than throwing into the app shell. */
export async function fetchCapabilities(
  apiBaseUrl: string,
  signal?: AbortSignal,
): Promise<CapabilityStatus[]> {
  const response = await authedFetch(
    `${apiBaseUrl}/api/capabilities`,
    signal ? { signal } : undefined,
  );
  if (!response.ok) return [];
  const body = await response.json();
  return Array.isArray(body) ? (body as CapabilityStatus[]) : [];
}

/** Whether the language model runs on its keyless fallback — i.e. this user's explanations are
 *  keyless too, and the per-device compute choice applies. */
export function isLlmKeyless(capabilities: CapabilityStatus[]): boolean {
  return capabilities.some((c) => c.capability === "llm" && c.mode === "fallback");
}
